docs/clients/promtail/stages/regex.md

```diff
@@ -88,3 +96,50 @@ and append the following key-value pairs back into the extracted map:
 
 - `year`: `2019`
 
+### With `replace`
+
+Given the pipeline:
+
+```yaml
+- regex:
+    expression: "password (?P<sensitve>\S+)"
+    replace: 
+      value: "****"
+      output: msg
+```
+
+And the log line:
+
+```
+2019-01-01T01:00:00.000000001Z info i'm a log message who has sensitive information with password xyz
+```
+
+The following key-value pairs would be added to the extracted map:
+
+- `sensitive`: `****`
+- `msg`: `2019-01-01T01:00:00.000000001Z info i'm a log message who has sensitive information with password ****`
+
+### With `replace` and value as template
+
+Given the pipeline:
+
+```yaml
+- regex:
+    expression: "\S+ level=(?P<level>\S+) msg=(?<msg>.*)"
```

Comment:
I'm sorry but seeing it now I prefer to have 2 different stage for different purpose. One to extract `regex` one to replace content `replace`.

The couple of cons with the current solution like what if I want to replace 2 entry ?
Also I might use json and replace an entry without the need to extract anything.

Reply:
I agree to make it two different stages. But the fields are the same except for the `replace` field.

> The couple of cons with the current solution like what if I want to replace 2 entry 

2 entries? Each entry will go through this pipeline stage and get replaced right?

I didn't get you completely. If I understood correctly `replace` is meant to replace some string in an entry right? So with the regex you match the string you want to replace

Reply:
2 fields sorry !

My main concern you should be able to replace without extracting data. I prefer to separate them mostly because it's easier to understand when concern are separated.

Reply:
> I prefer to separate them mostly because it's easier to understand when concern are separated.

Totally agree.

How do we plan to allow replacing 2 fields with single `replace` stage? the value given in `replace` field can be only one right? 

```
- regex:
    expression: "\S+ level=(?P<level>\S+) msg=(?<msg>.*)"
    replace: 
      value: {{ .Value | ToUpper }}
      output: log
```

All the captured group will be replaced with same content. Two support replacing 2 fields with 2 separate values we will have to allow adding a mapping like below

```
- regex:
    expression: "\S+ level=(?P<level>\S+) msg=(?<msg>.*)"
    replace: 
      level: {{ .Value | ToUpper }}
      msg: "somevalue"
```

Reply:
Yeah my was point it's weird cause you can extract 10 value but replace only one ? And this is why I think it's own replace stage would make sense. No need to support multiple values at once this was just an example.

Reply:
I like the idea to have separate `replace` stage

Just one more thing. Even with new `replace` stage it would be same right?

```
- replace:
    expression: "\S+ level=(?P<level>\S+) msg=(?<msg>.*)"
    source: "somesource"
    replace: {{ .Value | ToUpper }}
```

Even here user can extract one value or multiple values. There might be usecase where the user will replace two different fields with same value.

Reply:
yeah I don't think named match is required but doesn't matter, any match  should be replaced.

Replace should not extract data, just mutate extracted data.

```
- replace:
    expression: "\S+ level=(?P<level>\S+) msg=(?<msg>.*)"
    replacement: {{ .Value | ToUpper }}
```

Value should be the matched value and so this will cause a problem if you have multiple match
maybe we should use .$1 .$2 .$3 (or .Value[0]...) and named match .level not sure here, this could clash with extracted data.

Reply:
> maybe we should use .$1 .$2 .$3 (or .Value[0]...) and named match .level not sure here, this could clash with extracted data.

This is little tricky. Let me think on this

Reply:
Again not required for the first version, especially since you can extract and reuse after.

Does the template support for extracted data ?

Reply:
Yes, template supports extracted data